### PR TITLE
Add StreamInterceptor for Authenticating requests

### PR DESCRIPTION
**Pull request by PrasadG193**

### CODE REVIEW

File: proto/schema.proto [hunk 1]
```diff
@@ -10,22 +10,18 @@ service SnapshotMetadata {
...
 
 message GetDeltaRequest {
-	string security_token = 1;
-  	string namespace = 2;
-  	string volume = 3;
-  	string base_snapshot = 4;
-  	string target_snapshot = 5;
-  	uint64 starting_byte_offset = 6;
-  	uint32 max_results = 7;
+  	string volume = 1;
```
Comment (carlbraganza) [inline proto/schema.proto]: Drop `volume` too!

File: proto/schema.proto [hunk 2]
```diff
@@ -10,22 +10,18 @@ service SnapshotMetadata {
...
 }
 
 message GetAllocatedRequest {
-	string security_token = 1;
-	string namespace = 2;
-	string volume_id = 3;
-	string snapshot = 4;
-	uint64 starting_offset = 5;
-	uint32 max_results = 6;
+	string volume_id = 1;
```
Comment (carlbraganza) [inline proto/schema.proto]: `volume_id` not needed!

File: pkg/kube/utils.go [hunk 3]
```diff
@@ -35,17 +36,21 @@ func FindSnapshotMetadataService(ctx context.Context, cli client.Client, driver
...
 }
 
-func GetVolSnapshotInfo(ctx context.Context, cli client.Client, volSnapName, namespace string) (string, string, error) {
+func GetVolSnapshotInfo(ctx context.Context, cli client.Client, namespacedName string) (string, string, error) {
 	volSnap := &volsnapv1.VolumeSnapshot{}
-	err := cli.Get(ctx, types.NamespacedName{Name: volSnapName, Namespace: namespace}, volSnap)
+	vsName := strings.Split(namespacedName, "/")
+	if len(vsName) != 2 {
+		return "", "", fmt.Errorf("Invalid parameter. The snapshot names should be passed in namespace/name format.")
+	}
```
Comment (carlbraganza) [inline pkg/kube/utils.go:44]: 👍

File: cmd/server/main.go [hunk 4]
```diff
@@ -254,6 +279,14 @@ func loadTLSCredentials() (credentials.TransportCredentials, error) {
 	return credentials.NewTLS(config), nil
 }
 
+func findSnapshotNamespace(namespacedName string) (string, error) {
+	sp := strings.Split(namespacedName, "/")
+	if len(sp) != 2 {
+		return "", fmt.Errorf("Invalid parameter. The snapshot names should be passed in namespace/name format.")
+	}
+	return sp[0], nil
+}
```
Comment (carlbraganza) [inline cmd/server/main.go:315]: Do you use this anywhere?

A better signature would be `func ParseSnapshotID(snapID string) (ns, name string, err error)`.

File: cmd/server/main.go [hunk 5]
```diff
@@ -136,20 +142,45 @@ func (s *Server) getAudienceForDriver(ctx context.Context) (string, error) {
...
+	if len(securityToken) == 0 {
+		return status.Errorf(codes.Unauthenticated, "authorization token is not provided")
+	}
+
+	// TODO: Find better way to get namespace
+	ns := md["namespace"]
+	if len(ns) == 0 {
+		return status.Errorf(codes.Unauthenticated, "namespace is not provided")
+	}
+	namespace := ns[0]
```
Comment (carlbraganza) [inline cmd/server/main.go]: This should be from the snapshot_id field and not the metadata.

File: proto/schema.proto [hunk 6]
```diff
@@ -10,22 +10,18 @@ service SnapshotMetadata {
...
-	string security_token = 1;
-  	string namespace = 2;
-  	string volume = 3;
-  	string base_snapshot = 4;
-  	string target_snapshot = 5;
-  	uint64 starting_byte_offset = 6;
-  	uint32 max_results = 7;
+  	string volume = 1;
+  	string base_snapshot = 2;
+  	string target_snapshot = 3;
```
Comment (carlbraganza) [inline proto/schema.proto]: The CSI spec uses an `_id` suffix.

File: proto/schema.proto [hunk 7]
```diff
@@ -10,22 +10,18 @@ service SnapshotMetadata {
...
 
 message GetAllocatedRequest {
-	string security_token = 1;
-	string namespace = 2;
-	string volume_id = 3;
-	string snapshot = 4;
-	uint64 starting_offset = 5;
-	uint32 max_results = 6;
+	string volume_id = 1;
+	string snapshot = 2;
```
Comment (carlbraganza) [inline proto/schema.proto]: `snapshot_id`

File: cmd/client/main.go [hunk 8]
```diff
@@ -126,12 +127,33 @@ func (c *Client) createSAToken(ctx context.Context, audience string, sa, namespa
...
+		method string,
+		streamer grpc.Streamer,
+		opts ...grpc.CallOption,
+	) (grpc.ClientStream, error) {
+		return streamer(c.attachToken(ctx, token, namespace), desc, cc, method, opts...)
+	}
+}
+
+func (c *Client) attachToken(ctx context.Context, token, namespace string) context.Context {
+	return metadata.AppendToOutgoingContext(ctx, "authorization", token, "namespace", namespace)
```
Comment (carlbraganza) [inline cmd/client/main.go]: Shouldn't pass namespace via the metadata.

File: cmd/server/main.go [hunk 9]
```diff
@@ -136,20 +142,45 @@ func (s *Server) getAudienceForDriver(ctx context.Context) (string, error) {
 	return sms.Spec.Audience, nil
 }
 
-func (s *Server) authRequest(ctx context.Context, securityToken, namespace string) error {
+func (s *Server) streamInterceptor(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+	log.Println("Stream interceptor: performing auth checks for req", info.FullMethod)
+	if err := s.authRequest(ss.Context()); err != nil {
+		return err
+	}
```
Comment (carlbraganza) [inline cmd/server/main.go]: Should we also be checking if `info.IsServerStream` is true and/or that the names of the method match `GetDelta` or `GetAllocated`?  May be unnecessary and too complicated ....

File: cmd/server/main.go [hunk 10]
```diff
@@ -55,16 +58,18 @@ func main() {
...
 	if err != nil {
 		log.Fatal("cannot load TLS credentials: ", err)
 	}
-	s := grpc.NewServer(
+	opts := []grpc.ServerOption{
 		grpc.Creds(tlsCredentials),
-	)
-	reflection.Register(s)
-	pgrpc.RegisterSnapshotMetadataServer(s, newServer())
+		grpc.StreamInterceptor(srv.streamInterceptor),
```
Comment (carlbraganza) [inline cmd/server/main.go:72]: 👍

File: cmd/client/main.go [hunk 11]
```diff
@@ -126,12 +127,33 @@ func (c *Client) createSAToken(ctx context.Context, audience string, sa, namespa
...
+func (c *Client) initGRPCClient(cacert []byte, URL, token, namespace string) {
 	tlsCredentials, err := loadTLSCredentials(cacert)
 	if err != nil {
 		log.Fatal("cannot load TLS credentials: ", err)
 	}
-	conn, err := grpc.Dial(URL, grpc.WithTransportCredentials(tlsCredentials))
+	conn, err := grpc.Dial(
+		URL,
+		grpc.WithTransportCredentials(tlsCredentials),
+		grpc.WithStreamInterceptor(c.streamClientInterceptor(token, namespace)),
```
Comment (carlbraganza) [inline cmd/client/main.go]: Please try with `PerRPCCredentials` as it makes the most sense in this context.

File: cmd/client/main.go [hunk 12]
```diff
@@ -209,12 +231,10 @@ func (c *Client) getChangedBlocks(
...
 
-	c.initGRPCClient(snapMetaSvc.Spec.CACert, snapMetaSvc.Spec.Address)
+	c.initGRPCClient(snapMetaSvc.Spec.CACert, snapMetaSvc.Spec.Address, saToken, snapNamespace)
 	stream, err := c.client.GetDelta(ctx, &pgrpc.GetDeltaRequest{
-		SecurityToken:      saToken,
-		Namespace:          snapNamespace,
-		BaseSnapshot:       baseVolumeSnapshot,
-		TargetSnapshot:     targetVolumeSnapshot,
+		BaseSnapshot:       snapNamespace + "/" + baseVolumeSnapshot,
+		TargetSnapshot:     snapNamespace + "/" + targetVolumeSnapshot,
```
Comment (carlbraganza) [inline cmd/client/main.go]: 👍

File: cmd/client/main.go [hunk 13]
```diff
@@ -175,7 +197,7 @@ func (c *Client) setupSecurityAccess(
 	// 1. Find Driver name for the snapshot
 	fmt.Printf("\n## Discovering SnapshotMetadataService for the driver and creating SA Token \n\n")
 	log.Print("Finding driver name for the snapshots")
-	_, driver, err := kube.GetVolSnapshotInfo(ctx, c.rtCli, baseSnap, snapNamespace)
+	_, driver, err := kube.GetVolSnapshotInfo(ctx, c.rtCli, snapNamespace+"/"+baseSnap)
```
Comment (carlbraganza) [inline cmd/client/main.go:200]: No need to change this - it is internal

File: cmd/client/main.go (hunk 8 above)
Comment (PrasadG193) [reply cmd/client/main.go]: I have remove this

**Pull request merged**